File: Plan Graph (sgware)/src/edu/uno/ai/planning/pg/StepNode.java
package edu.uno.ai.planning.pg;

import java.util.ArrayList;
import java.util.Iterator;

import edu.uno.ai.planning.Step;
import edu.uno.ai.planning.logic.Literal;

public class StepNode extends Node {

	public final Step step;
	public final boolean persistence;
	protected final ArrayList<LiteralNode> preconditions = new ArrayList<>();
	protected final ArrayList<LiteralNode> effects = new ArrayList<>();
	private int literalCount = 0;
	
	protected StepNode(PlanGraph graph, Step step) {
		super(graph);
		this.step = step;
		this.persistence = false;
	}
	
	protected StepNode(PlanGraph graph, Literal literal) {
		super(graph);
		this.step = new Step("(persist " + literal + ")", literal, literal);
		this.persistence = true;
	}
	
	@Override
	public int hashCode() {
		return step.hashCode();
	}
	
	@Override
	public String toString() {
		return step.toString();
	}
	
	protected void incrementLiteralCount() {
		markForReset();
		literalCount++;
		if(literalCount == preconditions.size())
			graph.nextSteps.add(this);
	}
	
	@Override
	protected boolean setLevel(int level) {
		if(super.setLevel(level)) {
			for(LiteralNode effect : effects)
				effect.setLevel(level);
			return true;
		}
		else
			return false;
	}
	
	@Override
	protected void reset() {
		super.reset();
		literalCount = 0;
	}
	
	public Iterable<LiteralNode> getPreconditions(int level) {
		return new Iterable<LiteralNode>() {
			@Override
			public Iterator<LiteralNode> iterator() {
				return new NodeIterator<>(level - 1, preconditions);
			}
		};
	}
	
	public Iterable<LiteralNode> getEffects(int level) {
		return new Iterable<LiteralNode>() {
			@Override
			public Iterator<LiteralNode> iterator() {
				return new NodeIterator<>(level, effects);
			}
		};
	}
}

File: UnitySokoban/Assets/Scripts/Planning/PlanGraphSGW/StepNode.cs
using Planning;
using Planning.Logic;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace PlanGraphSGW
{
    public class StepNode : Node
    {
        public readonly Step step;
        public readonly bool persistence;
        internal readonly List<LiteralNode> preconditions = new List<LiteralNode>();
        internal readonly List<LiteralNode> effects = new List<LiteralNode>();
        private int literalCount = 0;

        public StepNode(PlanGraph graph, Step step)
            : base(graph)
        {
            this.step = step;
            this.persistence = false;
        }

        public StepNode(PlanGraph graph, Literal literal)
            : base(graph)
        {
            this.step = new Step("(persist " + literal + ")", literal, literal);
            this.persistence = true;
        }

        public override int GetHashCode()
        {
            return step.GetHashCode();
        }

        public override string ToString()
        {
            return step.ToString() + " " + level;
        }

        public void incrementLiteralCount()
        {
            markForReset();
            literalCount++;
            if (literalCount == preconditions.Count)
                graph.nextSteps.Add(this);
        }

        public override bool setLevel(int level)
        {
            if (base.setLevel(level))
            {
                foreach (LiteralNode effect in effects)
                    effect.setLevel(level);
                return true;
            }
            else
                return false;
        }

        internal override void Reset()
        {
            base.Reset();
            literalCount = 0;
        }

        public IEnumerable<LiteralNode> getPreconditions(int level)
        {
            IEnumerator<LiteralNode> iterator = new NodeIterator<LiteralNode>(level, preconditions);
            while (iterator.MoveNext())
                yield return iterator.Current;
        }

        public IEnumerable<LiteralNode> getEffects(int level)
        {
            IEnumerator<LiteralNode> iterator = new NodeIterator<LiteralNode>(level, effects);
            while (iterator.MoveNext())
                yield return iterator.Current;
        }
    }
}
